cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -298,54 +305,80 @@ public void run() {
-     * requested inter-broker partition reassignment concurrency.
+     * Initialize the reassignment concurrency adjustment with the load monitor and the initially requested reassignment concurrency.
      *
      * @param loadMonitor Load monitor.
      * @param requestedInterBrokerPartitionMovementConcurrency The maximum number of concurrent inter-broker partition movements
      *                                                         per broker(if null, use num.concurrent.partition.movements.per.broker).
+     * @param requestedLeadershipMovementConcurrency The maximum number of concurrent leader movements
+     *                                               (if null, use num.concurrent.leader.movements).
      */
-    public synchronized void initAdjustment(LoadMonitor loadMonitor, Integer requestedInterBrokerPartitionMovementConcurrency) {
+    public synchronized void initAdjustment(LoadMonitor loadMonitor,
+                                            Integer requestedInterBrokerPartitionMovementConcurrency,
+                                            Integer requestedLeadershipMovementConcurrency) {
       _loadMonitor = loadMonitor;
       setRequestedInterBrokerPartitionMovementConcurrency(requestedInterBrokerPartitionMovementConcurrency);
+      setRequestedLeadershipMovementConcurrency(requestedLeadershipMovementConcurrency);
     }
 
-    private boolean canRefreshConcurrency() {
-      return _concurrencyAdjusterEnabled && _executorState.state() == ExecutorState.State.INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS
-             && !_skipAutoRefreshingConcurrency && _loadMonitor != null;
+    private boolean canRefreshConcurrency(ConcurrencyType concurrencyType) {
+      if (!_concurrencyAdjusterEnabled.get(concurrencyType) || _loadMonitor == null) {
+        return false;
+      }
+      switch (concurrencyType) {
+        case LEADERSHIP:
+          return _executorState.state() == LEADER_MOVEMENT_TASK_IN_PROGRESS;
+        case INTER_BROKER_REPLICA:
+          return _executorState.state() == ExecutorState.State.INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS
+                 && !_skipInterBrokerReplicaConcurrencyAdjustment;
+        default:
+          throw new IllegalArgumentException("Unsupported concurrency type " + concurrencyType + " is provided.");
+      }
     }
 
-    private synchronized void refreshConcurrency() {
-      if (canRefreshConcurrency()) {
-        Integer recommendedConcurrency = ExecutionUtils.recommendedConcurrency(_loadMonitor.currentBrokerMetricValues(),
-                                                                               _executionTaskManager.interBrokerPartitionMovementConcurrency(),
-                                                                               _maxPartitionMovementsPerBroker);
+    private synchronized void refreshInterBrokerReplicaConcurrency() {
+      if (canRefreshConcurrency(ConcurrencyType.INTER_BROKER_REPLICA)) {
+        Integer recommendedConcurrency
```

Comment:
Nit: I think this statement (353 and 354) and the next line (355) can be swapped. Same case for line 365, 366 and 367

Reply:
Not sure I follow the recommendation here -- do you mean _shift the RHS of assignment up to the same line as its LHS_?

Reply:
I see. Nvm. I did not notice that the result of `_executionTaskManager.interBrokerPartitionMovementConcurrency()` is one of the input parameters passed to the `recommendedConcurrency ` method. Sorry for the confusion.